### ISSUE TITLE
Dispose WebGL context and web workers.

### ISSUE THREAD

**Issue by Jule-** [WhitestormJS/whitestorm.js#152]
Hi,
I am working on an SPA and I have to instanciate multiple times on different views the same component with a WebGL context created with `WHS.World`.

My issue is that when my view is detached you give no way to dispose the scene and the WebGL context created. Changing views leaks web workers and GPU memory. I have tried to terminate physic web worker but GPU memory is still massively leaking.
```javascript
this.world.$scene._worker.terminate()
```
This is a blocking issue for me. 😢

I use v1.2.6 but for what I've seen there is nothing more in v2.x.x that can fix my issue.

Is there something that I've missed in your code or is there some workaround that I can apply?
At least are you planing some API that can dispose context properly?

Thanks.

------

###### Version:
  - [x] v2.x.x
- [x] v1.x.x

###### Issue type:
 - [x] Bug
 - [ ] Proposal/Enhancement
 - [x] Question

------

<details> 
<summary> <b>Tested on: </b> </summary>

###### --- Desktop
   - [x] Chrome
   - [ ] Chrome Canary
   - [ ] Chrome dev-channel
   - [ ] Firefox
   - [ ] Opera
   - [ ] Microsoft IE
   - [ ] Microsoft Edge

###### --- Android
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Opera

###### --- IOS
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Opera

</details>

**Comment by sasha240100** [comment]
@Jule I see that you use v1 which is deprecated now. That's not a problem cause migration to v2 is easy and won't take much time.

There is a specific reason of why strongly suggest such migration. V2 is simply much more flexible and up-to-date.

in v2 each part of framework is replacable. As from what I understand - you can patch/overwrite WorldModule.  I will soon add API that can stop Physics simulation when needed. 

I am always online in discord channel and ready to help you

**Comment by sasha240100** [comment]
@Jule- I fixed it in https://github.com/WhitestormJS/physics-module-ammonext/commit/203350898896fd581fd3a1b345bd6f8cb013900b now. (on npm: v0.1.1)
That's how you can do that:
In v2.x.x we have everything which executes each frame wrapped with `WHS.Loop` class.

`WorldModule` now contains `simulationLoop` which you can stop/restart. That's how:

```javascript
const worldModule = new PHYSICS.WorldModule();

const app = new WHS.App([
  // ...
  worldModule
]);

worldModule.simulationLoop.stop(); // Stop simulation (terminate)
worldModule.simulationLoop.start(); // Start/play simulation

```

Same is also possible with `RenderingModule` which has `.renderLoop`. Can i close this issue now?

**Comment by Jule-** [comment]
I need to get it a try but I have to migrate my code first...

Nevertheless I've took a look to your fix and I have some suspections about the worker you create. I don't think stopping the loop and removing it from World can stop your worker and GC wont be able to free memory.

Like I said my issue is not really a matter of performance but more a matter of graphical memory freeing. I don't know what is really handling graphical context in your lib right now and prevent GC from releasing it.

I'll come back when I have tried v2. Thank you for the quick response! ;)

**Comment by Jule-** [comment]
If I can I'll make you a test case.

**Comment by Jule-** [comment]
And there it is: https://github.com/Jule-/AureliaWHS

**Comment by sasha240100** [comment]
@Jule-  I discovered this topic a bit. That's not, actually, a problem of whs, it is more related to Three.js. WHS just covers main parts of three.js such as it's scene, cameras and meshes (etc.).

**How to fix memory leak**

I researched some posts on stackoverflow. `geometry` as well as `material` - both have `.dispose()` method which removes geometry/material instance from renderer context.

- [`.dispose()`](https://threejs.org/docs/index.html?q=geometry#Reference/Core/Geometry.dispose) [Geometry]
- [`.dispose()`](https://threejs.org/docs/index.html?q=mater#Reference/Materials/Material.dispose) [Material]

Best way to achieve the expected result with the proper way is to write a [module](https://whsjs.readme.io/docs/modules-1) which will handle executing dispose on all scene's children when needed.

**Refs**

[GitHub]

https://github.com/mrdoob/three.js/issues/5175

[Stackoverflow]

- [Memory leak with three.js and many shapes](http://stackoverflow.com/questions/20997669/memory-leak-in-three-js)
- [How to remove and dispose of all child geometry and meshes from an Object3D?](http://stackoverflow.com/questions/28723939/how-to-remove-and-dispose-of-all-child-geometry-and-meshes-from-an-object3d)
- [Three js memory management](http://stackoverflow.com/questions/13914959/three-js-memory-management)

**Comment by phas0r** [comment]
There is example code in the three.js repository on how to dispose:
https://github.com/mrdoob/three.js/blob/master/examples/webgl_test_memory.html

**Comment by sasha240100** [comment]
@phas0r, @Jule- wants to remove the scene completely (with all reserved memory).

**Comment by sasha240100** [comment]
Can I close it?
/ping @Jule-

**Comment by Jule-** [comment]
Ok. I have found how to recover GPU memory when I want to! 🥇

In fact the documentation is ambiguous but it seems that
```javascript
gl.getExtension('WEBGL_lose_context').loseContext();
```
is working in Chrome and it is not just a placebo for testing purpose!

And I don't know why I haven't see it before but I this method seems to do the job too:
```javascript
app.$rendering.forceContextLoss();
```

That is the nice part of the story, unfortunately I've found some memory leaks with WHS I think:
![image](https://cloud.githubusercontent.com/assets/9716736/23670145/177f4fc6-0367-11e7-971d-6e8635950666.png)

- Snapshot 1: before initializing the scene (before `new WHS.App()`)
- Snapshot 2: after initializing the scene
- Snapshot 3: after releasing the gl context and all WHS variables.

For what I understand you are making references to DOM Div and/or Canvas inside WHS context or sub-context and never releasing them. I have not tried with non minified WHS so the source localisation is pretty hard to figure.
